src/algebra/module/zlattice.lean

```diff
@@ -0,0 +1,225 @@
+  fract b x = x ↔ x ∈ fundamental_domain b :=
+by simp only [basis.ext_elem_iff b, repr_fract_apply, int.fract_eq_self, mem_fundamental_domain,
+  set.mem_Ico]
+
+variable (b)
+
+lemma fract_mem_fundamental_domain (x : E) :
+  fract b x ∈ fundamental_domain b := fract_eq_self.mp (fract_fract b _)
+
+lemma fract_eq_fract (m n : E) :
+  fract b m = fract b n ↔ -m + n ∈ span ℤ (set.range b) :=
+begin
+  rw [eq_comm, basis.ext_elem_iff b],
+  simp_rw [repr_fract_apply, int.fract_eq_fract, eq_comm, basis.mem_span_iff_repr_mem,
+    sub_eq_neg_add, map_add, linear_equiv.map_neg, finsupp.coe_add, finsupp.coe_neg, pi.add_apply,
+    pi.neg_apply, ← (eq_int_cast (algebra_map ℤ K) _), set.mem_range],
+end
+
+lemma norm_fract_le (m : E) :
+  ‖fract b m‖ ≤ ∑ i, ‖b i‖ :=
+begin
+  calc
+    ‖fract b m‖ = ‖∑ i, b.repr (fract b m) i • b i‖ : by rw b.sum_repr
+            ... = ‖∑ i, int.fract (b.repr m i) • b i‖ : by simp_rw repr_fract_apply
+            ... ≤ ∑ i, ‖int.fract (b.repr m i) • b i‖ : norm_sum_le _ _
+            ... ≤ ∑ i, ‖int.fract (b.repr m i)‖ * ‖b i‖ : by simp_rw norm_smul
+            ... ≤ ∑ i, ‖b i‖ : finset.sum_le_sum (λ i _, _),
+    suffices : ‖int.fract (((b.repr) m) i)‖ ≤ 1,
+    { convert mul_le_mul_of_nonneg_right this (norm_nonneg _ : 0 ≤ ‖b i ‖),
+      exact (one_mul _).symm, },
+    rw (norm_one.symm : 1 = ‖(1 : K)‖),
+    apply norm_le_norm_of_abs_le_abs,
+    rw [abs_one, int.abs_fract],
+    exact le_of_lt (int.fract_lt_one _),
+end
+
+section unique
+
+variable [unique ι]
+
+@[simp] lemma coe_floor_self (k : K) : (floor (basis.singleton ι K) k : K) = ⌊k⌋ :=
+basis.ext_elem _ (λ _, by rw [repr_floor_apply, basis.singleton_repr, basis.singleton_repr])
+
+@[simp] lemma coe_fract_self (k : K) : (fract (basis.singleton ι K) k : K) = int.fract k :=
+basis.ext_elem _ (λ _, by rw [repr_fract_apply, basis.singleton_repr, basis.singleton_repr])
+
+end unique
+
+end fintype
+
+lemma fundamental_domain_bounded [finite ι] :
+  metric.bounded (fundamental_domain b) :=
+begin
+  casesI nonempty_fintype ι,
+  use 2 * ∑ j, ‖b j‖,
+  intros x hx y hy,
+  refine le_trans (dist_le_norm_add_norm x y) _,
+  rw [← fract_eq_self.mpr hx, ← fract_eq_self.mpr hy],
+  refine (add_le_add (norm_fract_le b x) (norm_fract_le b y)).trans _,
+  rw ← two_mul,
+end
+
+lemma exist_unique_vadd_mem_fundamental_domain [finite ι] (x : E) :
+  ∃! v : span ℤ (set.range b), v +ᵥ x ∈ fundamental_domain b :=
+begin
+  casesI nonempty_fintype ι,
+  refine ⟨-floor b x, _, λ y _, _⟩,
+  { simp_rw [fundamental_domain, set.mem_Ico, vadd_def, vadd_eq_add, add_subgroup_class.coe_neg,
+    neg_add_eq_sub, ← fract_apply],
+    simp only [repr_fract_apply, int.fract_nonneg, int.fract_lt_one, true_and, set.mem_set_of_eq,
+      implies_true_iff], },
+  { rwa [subtype.ext_iff, ← add_right_inj x, add_subgroup_class.coe_neg, ← sub_eq_add_neg,
+      ← fract_apply, ← fract_zspan_add b _ (subtype.mem y), add_comm, ← vadd_eq_add, ← vadd_def,
+      eq_comm, fract_eq_self], },
```

Comment:
These two goals look like they might be good standalone lemmas.

Reply:
I added 
```lean
lemma vadd_mem_fundamental_domain [fintype ι] (y : span ℤ (set.range b)) (x : E) :
  y +ᵥ x ∈ fundamental_domain b ↔ y = -floor b x 
```
And it does make things simpler :)